Taproot swaps (V2)

JssDWt opened the pull request:
This is an alternative to #1176

There is a single interface to swaps: `BTCReceiveSwap`. The main logic for swap execution is inside this struct, with some divergent paths for either segwit and taproot swaps where applicable. Most of the logic for the segwit and taproot swaps are the same, so they can be bundled together like this.

The data model is the same as before, only additional chain information has been added to the SwapInfo table. This is mainly because the current data model only considered utxos, but it's useful to also consider spent outputs in circumstances where multiple outputs exist for the same swap address.

The invoice creation logic is different from the original version. Invoices are deleted when the amount no longer matches, or when the local liquidity no longer matches. A missing piece of the puzzle there is to allow re-registration of a payment in lspd.

roeierez commented on libs/sdk-core/src/swap_in/swap.rs:
```diff
@@ -194,384 +234,283 @@ impl BTCReceiveSwap {
-        Ok(self.persister.list_swaps(ListSwapsRequest {
-            status: Some(SwapStatus::refundable()),
+        })?;
+        let recent = self.list_swaps(ListSwapsRequest {
```
So we are rescanning anyway all swaps until timeout? regardless of their status?  Just curious what have we been missing before this implementation.

JssDWt replied:
This is just to make sure we catch additional funds sent to the same address within a reasonable time.

roeierez reviewed (commented):
Looks really good! dropped some small comments and I still want to review the state resolution as it has caused issues in the past.

JssDWt commented on libs/sdk-core/src/greenlight/node_api.rs:
```diff
@@ -1065,6 +1066,35 @@ impl NodeAPI for Greenlight {
+
+        let status = match result.status() {
+            ListinvoicesInvoicesStatus::Unpaid => DelinvoiceStatus::Unpaid,
+            ListinvoicesInvoicesStatus::Paid => DelinvoiceStatus::Paid,
```
Right! Done in 3a9f1a430ebd3f323a31d36eed5068684dc1a27d

JssDWt commented on libs/sdk-core/src/persist/swap.rs:
```diff
@@ -442,214 +483,3 @@ impl SqliteStorage {
     }
 }
-
-#[cfg(test)]
```
I will write new tests. It's the tests for the state transitions that no longer work, because state transitions are now handled inside `BTCReceiveSwap`, rather than on `SwapInfo` itself.

danielgranhao commented on libs/sdk-core/src/swap_in/segwit.rs:
```diff
@@ -0,0 +1,178 @@
+
+    pub async fn get_swap_payment(
+        &self,
+        _swap_info: &SwapInfo,
```
Nit: can we get rid of this parameter?

danielgranhao commented on libs/sdk-core/src/swap_in/segwit.rs:
```diff
@@ -0,0 +1,178 @@
+        Ok(tx)
+    }
+
+    pub async fn get_swap_payment(
```
Nit: this name makes it seem like we're getting some "payment" struct (like reading from db). Maybe `complete_swap_payment` or `complete_payment`?

JssDWt replied:
I went for `payout_swap`: 9fec848ab67f58da6bf6ec7678355509e28ab60c

danielgranhao reviewed (commented):
Did a first pass and looks good! Left some minor comments

JssDWt commented on libs/sdk-core/src/persist/transactions.rs:
```diff
@@ -317,6 +317,7 @@ impl SqliteStorage {
     }
 
     /// Look up a modified open channel bolt11 by hash.
+    #[allow(unused)]
     pub(crate) fn get_open_channel_bolt11_by_hash(
```
It's still used in a test. Also I felt it should still be used. I'll look at where I can get the open channel bolt11 from the database instead of recreating it. Otherwise I might remove that functionality.

JssDWt commented on libs/sdk-core/src/swap_in/swap.rs:
```diff
@@ -581,892 +520,641 @@ impl BTCReceiveSwap {
-        let address = crate::bitcoin::Address::p2wsh(&script, crate::bitcoin::Network::Bitcoin);
-        let address_str = address.to_string();
-        assert_eq!(address_str, expected_address);
+        if refundable_utxos.is_empty() {
+            return SwapStatus::Completed;
+        }
```
Looking at the current implementation, there's 2 options for that scenario:
- if the timelock passed for the first confirmed output, it returns 'Completed'
- if the timelock has not yet passed for the first confirmed output, it returns 'Redeemed'

I think Redeemed is the right option here, what do you think?

JssDWt commented on libs/sdk-core/src/swap_in/segwit.rs:
```diff
@@ -0,0 +1,178 @@
+                        witness: Witness::from_vec(vec![
+                            [1; 73].to_vec(),
+                            Vec::new(),
+                            [1; 100].to_vec(),
```
Done in f247c006b595eecf0c324b3e72ba507e799ba31d

JssDWt commented on libs/sdk-core/src/swap_in/segwit.rs:
```diff
@@ -0,0 +1,178 @@
+
+    pub fn create_fake_refund_tx(
+        &self,
+        _swap_info: &SwapInfo,
```
Done in 4d90fae76c825d677d009a4c3b2f2d53513d70c3

roeierez commented on libs/sdk-core/src/swap_in/swap.rs:
```diff
@@ -581,892 +545,1100 @@ impl BTCReceiveSwap {
-mod tests {
-    use std::time::{SystemTime, UNIX_EPOCH};
-    use std::{sync::Arc, vec};
+    fn calculate_status_without_chain_data(
```
In case use sent above the limit, should we consider the swap as refundable immediately? Currently it seems to be refundable only after the timeout?

JssDWt replied:
That makes sense for taproot swaps, yes. Will do.

JssDWt replied:
Done in fdb402b4538b269ece53ed115e6748b4d8e808d6

roeierez commented on libs/sdk-core/src/swap_in/swap.rs:
```diff
@@ -581,892 +545,1100 @@ impl BTCReceiveSwap {
-}
+/// ReceiveSwapper private functions
+impl BTCReceiveSwap {
+    fn calculate_status(
```
Very Nice! I went over the logic several times and didn't find any edge cases.

roeierez reviewed (commented):
Looks really good. Some small comments.

JssDWt commented:
Things fixed in the last few days:
- the `script` parameter is set to the refund script. In fact this is not enough to spend. It should probably be both scripts. I didn't yet find a good way to serialize the entire taptree.
- added some tests
- Fixed the cooperative refund transaction creation
- Fixed fee calculation in prepare_refund
- Return a proper error when utxos are still timelocked in the unilateral case
- Ensure refund transactions can be RBF'ed
- Ensure the status is still `Refundable` when the refund tx is unconfirmed
